in_toto/models/link.py

```diff
@@ -85,22 +91,58 @@ class Link(Signable):
     _type = attr.ib()
     name = attr.ib()
     materials = attr.ib()
-    products = attr.ib()
     byproducts = attr.ib()
     command = attr.ib()
     environment = attr.ib()
 
     def __init__(self, **kwargs):
         super().__init__()
 
-        self._type = "link"
-        self.name = kwargs.get("name")
-        self.materials = kwargs.get("materials", {})
-        self.products = kwargs.get("products", {})
-        self.byproducts = kwargs.get("byproducts", {})
-        self.command = kwargs.get("command", [])
-        self.environment = kwargs.get("environment", {})
+        link = linkpb.Link()
+        link.name = kwargs.get("name", "")
+        link.byproducts.update(kwargs.get("byproducts", {}))
+        link.environment.update(kwargs.get("environment", {}))
+
+        for material in kwargs.get("materials", []):
+            if not isinstance(material, rdpb.ResourceDescriptor):
+                material = rdpb.ResourceDescriptor(**material)
+
+            link.materials.append(material)
+
+        for command in kwargs.get("command", []):
+            link.command.append(command)
+
+        # NOTE: We need to serialize the protobuf into JSON
+        # so that we can sign the contents downstream
+        # serialized_link = MessageToDict(link)
+        serialized_link = MessageToJson(message=link)
+        serialized_link = json.loads(serialized_link)
+
+        # This is to assign attributes onto the class
+        # itself so the properties are directly accessible.
+        for k, v in serialized_link.items():
+            setattr(self, k, v)
+
+        if hasattr(self, 'name') == False:
+            self.name = ""
+
+        if hasattr(self, 'materials') == False:
+            self.materials = []
 
+        if hasattr(self, 'command') == False:
+            self.command = []
+
+        if hasattr(self, 'environment') == False:
+            self.environment = {}
+
+        if hasattr(self, 'byproducts') == True:
+            self.byproducts["return-value"] = int(
```

Comment:
@adityasaky In order to successfully serialize this class into a canonical JSON, I've had to write in this hack. When the class is parsed from a protobuf to a JSON, the value becomes a float, which breaks downstream processes because that is [against the rules of canonical JSON](https://github.com/secure-systems-lab/securesystemslib/blob/main/securesystemslib/formats.py#L684). In the current design, we cannot get around this limitation because it is a central part of the `Signable` class and its `signable_bytes` property. 

As you may have already witnessed, either due to the architecture or my poor Python skills (most likely the latter), this class (and the Statement class) became quite ugly when trying to integrate the protobufs into the existing structure. As the code is written now, I believe there's not a way around using the Signable class as it is integral in generating and validating signatures. I fear a slight refactor might be in the near future, but I wanted your thoughts first.

Any thoughts on the best way to proceed?

cc: @marcelamelara

Reply:
If I'm not mistaken, the link predicate shouldn't be `Signable` at all. The predicate will be embedded in a Statement which is what we'd actually be signing.

Reply:
I missed that the link was extending `Signable`. Agree with Aditya that shouldn't be the case. The link proto should be converted into a struct, inserted into the Statement `predicate` field, and the Statement will then be signed. 

I also believe that Statement might not have to extend `Signable` anymore either. The validation of the structure is now implemented in the Statement wrapper class provided by the `in_toto_attestation` package, and we can similarly use protobuf APIs directly to expose the signable bytes for a Statement.